addons/cloud_storage_google/models/cloud_storage_google.py

```diff
@@ -0,0 +1,92 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import json
+import re
+from urllib.parse import unquote, quote
+
+try:
+    from google.oauth2 import service_account
+    from google.auth.transport.requests import Request
+except ImportError:
+    service_account = Request = None
+
+from odoo import models
+from odoo.exceptions import ValidationError
+
+from ..utils.cloud_storage_google_utils import generate_signed_url_v4
+
+GOOGLE_CLOUD_STORAGE_ENDPOINT = 'https://storage.googleapis.com'
+
+
+class CloudStorageGoogle:
+    _url_pattern = re.compile(rf'{GOOGLE_CLOUD_STORAGE_ENDPOINT}/(?P<bucket_name>[\w\-.]+)/(?P<blob_name>[^?]+)')
+    _upload_url_time_to_expiry = 300  # 300 seconds
+    _download_url_time_to_expiry = 300  # 300 seconds
+    __credentials = {}  # {db_name: (account_info, credential)}
+
+    @classmethod
+    def get_info_from_url(cls, url):
+        match = cls._url_pattern.match(url)
+        if not match:
+            raise ValidationError('%s is not a valid Google Cloud Storage URL.', url)
+        return {
+            'bucket_name': match['bucket_name'],
+            'blob_name': unquote(match['blob_name']),
+        }
+
+    @classmethod
+    def get_credential(cls, env):
+        """ Get the credentials object of currently used account info.
+        This method is cached to because from_service_account_info is slow.
+        """
+        cached_account_info, cached_credential = cls.__credentials.get(env.registry.db_name, (None, None))
+        account_info = json.loads(env['ir.config_parameter'].get_param('cloud_storage_google_account_info'))
+        if cached_account_info == account_info:
+            return cached_credential
+        credential = service_account.Credentials.from_service_account_info(account_info)
+        cls.__credentials[env.registry.db_name] = (account_info, credential)
+        return credential
+
+    @classmethod
+    def generate_url(cls, env, blob_name):
```

Comment:
Same remark as above.  All those class methods look like model methods that ended up in the wrong class.